kernel/lru.c

```diff
@@ -0,0 +1,102 @@
+#include "types.h"
+#include "riscv.h"
+#include "param.h"
+#include "stat.h"
+#include "spinlock.h"
+#include "defs.h"
+#include "proc.h"
+#include "sleeplock.h"
+#include "fs.h"
+#include "buf.h"
+#include "file.h"
+#include "vm.h"
+#include "lru.h"
+
+
+
+
+
+// Initialize global pointers for the LRU list
+struct page_info *lru_head = 0;  // Head of the LRU list
+struct page_info *lru_tail = 0;  // Tail of the LRU list
+
+// Insert a page at the head of the LRU list
+void insert_into_lru(struct page_info *page) {
+    if (lru_head) {
+        lru_head->prev = page;
+    }
+    page->next = lru_head;
+    page->prev = 0;
+    lru_head = page;
+
+    // If the list was empty, the new page becomes both the head and the tail
+    if (lru_tail == 0) {
+        lru_tail = page;
+    }
+}
+
+// Remove a page from the LRU list
+void remove_from_lru(struct page_info *page) {
+    // Adjust pointers to remove the page from the list
+    if (page->prev) {
+        page->prev->next = page->next;
+    } else {
+        lru_head = page->next;
+    }
+    if (page->next) {
+        page->next->prev = page->prev;
+    } else {
+        lru_tail = page->prev;
+    }
+    page->next = 0;
+    page->prev = 0;
+}
+
+// Optional: Print the LRU list for debugging
+void print_lru_list(void) {
+    struct page_info *p = lru_head;
+    while (p) {
+        printf("Page: %p\n", p);
+        p = p->next;
+    }
+}
+
+
+
+void evict_page(void) {
+    printf("Evicting page...\n");
+    // Select the Least Recently Used (LRU) page (tail of the global LRU list)
+    struct page_info *victim = lru_tail;
+  
+    // If there is no page to evict, return
+    if (victim == NULL) {
+      return;  // No page to evict
+    }
+  
+    // Remove the page from the global LRU list
+    remove_from_lru(victim);
+  
+    // Write the page to swap space (if necessary)
+    if (victim->in_swap) {
+      swap_out_page(victim);  // Swap out the page if it was in swap
+    }
```

Comment:
A page in swap can't be swapped out.
Either your check is wrong, or your correct variable names.